_source/_posts/2020-12-08-beginners-guide-to-jwt.adoc

```diff
@@ -0,0 +1,175 @@
+
+WARNING: When using private claims, watch out for name collisions with the official claims.
+
+The use of registered claims is optional, but when they are present, they MUST be validated.  For example, a JWT may contain date-time fields that describe when the token is valid.
+
+- Issued At (`iat`) - The time the JWT was created
+- Expiration Time (`exp`) - The time at which the JWT is no longer valid
+- Not Before (`nfb`) - The earliest time the JWT would be valid
+
+NOTE: Timestamps are "seconds since the epoch" integer format. JWT libraries can add up to a few minutes of leeway to these values to account for clock skew between systems.
+
+=== JWT Header
+
+The header of a JWT contains information about how the token was created.  In my example, the "algorithm" (`alg`) claim is set to `HS256`, which specifies the hashing algorithm HMAC SHA-256 is used to generate or validate the signature.
+
+[source,json]
+----
+{
+  "alg": "HS256"
+}
+----
+
+=== JWT Signature
+
+The JWT specifications list a few different signing algorithms; each of these algorithms work slightly different. For simplicity's sake, there are two types of algorithms:
+- HMAC based shared secret, these all start with the prefix `HS`, which stands for HMAC SHA)
+- Public key pair (either RSA or ECDSA keys)
+
+WARNING: Caution is needed when using a shared secret, as anyone with the secret can create (or forge) new JWTs. If you need to validate a JWT from an untrusted client (web-page, mobile app, etc.), use a public key pair instead.
+
+The JWT in this example (actually a JWS, remember the `S` stands for "signature") uses the `HS256` algorithm.  To validate the signature, the hash of the first two parts of the token is rehashed, and then output it as a base64-url string.
+
+On the command line, you can use `openssl` to check the signature:
+
+[source,sh]
+----
+echo -n 'eyJhbGciOiJIUzI1NiJ9.eyJuYW1lIjoiSm9lIENvZGVyIn0' | \
+openssl dgst -sha256 -macopt hexkey:${secret_key_in_hex} -mac hmac -binary | \
+gbasenc --base64url | sed 's/=//'
+----
+
+If the output matches the original signature block, the signature is valid.
+
+== Problems with JWTs
+
+Fully validating a JWT is MUCH more complex than running a couple CLI commands. There are many edge cases and exploits; you should **ALWAYS** use a trusted JWT library and keep it up to date.
+
+////
+TIP: Visit https://token.dev[token.dev] to debug JWTs from within your browser!
+////
+
+One of the biggest problems with the JWT, is the signature verification to be disabled by setting the algorithm header claim to `none`. Many JWT library vulnerabilities have been related to the `none` algorithm.
+
+[source,txt]
+----
+eyJhbGciOiJub25lIn0.eyJuYW1lIjoiSm9lIENvZGVyIn0.
+----
+This decodes to:
+
+[source,json]
+----
+{"alg":"none"}
+.
+{"name":"Joe Coder"}
+----
+
+There is nothing secure about two JSON strings separated by a period.
```

Comment:
Did you intend to expand on this? If so, I suggest doing so above where you explain JWS v. JWE and show examples with base64 decoding of the payloads.